Use new heartbeat package

leogermani opened the pull request:
This PR makes jetpack uses the newly created Heartbeat package to handle the Heartbeat. It also creates a new CLI command to inspect the heartbeat

The approach here was to touch as little as possible the current heartbeat class, so everything still works as usual. `Jetpack_Heartbeat` class is now acting as a proxy class to the package, so you still can initialize and call its methods.

**Important** - There are some uses of the `generate_stats_array` method that are not related to the heartbeat itself. That's why we kept it as is and static. It's used, for example, by `wp jetpack status` CLI command.

That's why these stats are always returned by the static method. On the other hand, they are only added to the heartbeat stats batch if the connection is active. (Originally, the whole heartbeat was only initialized if the connection was active)

This PR depends on the new Heartbeat package -> #16276

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
* Uses Heartbeat package to handle heartbeat

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* 1146297891914257-as-1146297891914257
* p9dueE-1tz-p2

#### Does this pull request change what data or activity we track or use?
no

#### Testing instructions:
* Nothing should change and heartbeat should work as usual, you have to manually debug the heartbeat to see if it's still working by adding debug functions in its `cron_exec`method.

Here's how to manually trigger the heart beat:
* `wp jetpack options update last_heartbeat 0` - to trick Jetpack into running the heartbeat on command.
* `wp cron event run jetpack_v2_heartbeat` - to run the heartbeat on command.

* Test the WP CLI command: `wp jetpack-heartbeat` and confirm it outputs all the blog details that are sent via heartbeat

sergeymitr commented on packages/heartbeat/src/class-heartbeat.php:
```diff
@@ -0,0 +1,244 @@
+		add_filter( 'jetpack_xmlrpc_methods', array( __CLASS__, 'jetpack_xmlrpc_methods' ) );
+
+		if ( defined( 'WP_CLI' ) && WP_CLI ) {
+			\WP_CLI::add_command( 'jetpack-heartbeat', array( $this, 'cli_callback' ) );
```
This makes the command `wp jetpack-heartbeat`, so I guess either the code of the PR description need to change 🙂

sergeymitr commented on packages/heartbeat/src/class-heartbeat.php:
```diff
@@ -0,0 +1,244 @@
+
+		\WP_CLI\Utils\format_items( 'table', $formatted_stats, array( 'Stat name', 'Bin' ) );
+
+		$last_heartbeat = \Jetpack_Options::get_option( 'last_heartbeat' );
```
Do you think it would make sense to import `Jetpack_Options` and `WP_CLI` and get rid of the excessive `\` throughout the file?

sergeymitr reviewed (changes requested):
Everything looks good and works fine as far as I can tell.

leogermani committed:
add heartbeat as a dependecy for jetpack

leogermani committed:
Replace internal calls to use Heartbeat package

leogermani committed:
do not deactivate the heartbeat upon disconnect

We are only adding jetpack stats to the heartbeat when jetpack is
connected

We dont need, and dont want, to disable all heartbeat when disconnecting
since that may affect other plugins/packages using Heartbeat

leogermani committed:
adds heartbeat wp cli command

leogermani committed:
bump version in docblocks

leogermani commented:
@sergeymitr thanks for the review

I've rebased the branch and applied the changes you suggested. (I personally don't have a problem with `\`, I like them, but I understand...)

The cli command must be `jetpack-heartbeat` with a dash so it works even when jetpack is not present

sergeymitr reviewed (dismissed):
LGTM 🙂

sergeymitr commented:
It looks like more reviews are needed, so I'm keeping the "Needs Review" label.

jeherve commented on class.jetpack.php:
```diff
@@ -3310,8 +3310,6 @@ public static function disconnect( $update_activated_state = true ) {
 		// Delete all the sync related data. Since it could be taking up space.
 		Sender::get_instance()->uninstall();
 
-		// Disable the Heartbeat cron
-		Jetpack_Heartbeat::init()->deactivate();
```
That's a bit tricky. I get why we can't just deactivate all cron jobs as it may impact other plugins using this package, but in this case how can we make sure we don't leave active scheduled events laying around? Am I missing something?

leogermani replied:
Here

https://github.com/Automattic/jetpack/pull/16285/files#diff-5d5dd765ee9393dec92c944543326154R175

We are only adding Jetpack info to the heartbeat if the connection is active. This is called the moment the data will be sent. So even if the heartbeat action triggers, no information will be sent.

In other words, the Heartbeat action will always trigger, but chances are it will have no data to send and will simply do nothing.

Does it answer your question?

jeherve replied:
It does, thank you. 👍

jeherve commented on class.jetpack-heartbeat.php:
```diff
@@ -162,27 +122,63 @@ public static function generate_stats_array( $prefix = '' ) {
-			wp_unschedule_event( $timestamp, 'jetpack_heartbeat' );
+		// Cronjobs are now handled by the Heartbeat package and we don't want to deactivate it here.
+		// We are adding jetpack stats to the heartbeat only if the connection is available. so we don't need to disable the cron when disconnecting.
+		_deprecated_function( __METHOD__, 'jetpack-8.8.0', 'Automattic\\Jetpack\\Heartbeat::deactivate' );
```
```suggestion
		_deprecated_function( __METHOD__, 'jetpack-8.9.0', 'Automattic\\Jetpack\\Heartbeat::deactivate' );
```

jeherve commented on class.jetpack-heartbeat.php:
```diff
@@ -162,27 +122,63 @@ public static function generate_stats_array( $prefix = '' ) {
 	public static function jetpack_xmlrpc_methods( $methods ) {
-		$methods['jetpack.getHeartbeatData'] = array( __CLASS__, 'xmlrpc_data_response' );
-		return $methods;
+		return Heartbeat::jetpack_xmlrpc_methods( $methods );
```
Should we add a deprecation notice here and below, before we return the new method?

leogermani replied:
Yes, I think in those 2 cases we are fine. They are not used anywhere else

jeherve reviewed (commented):
I have some questions as I am not sure I understand the inner workings here.

leogermani committed:
Update class.jetpack-heartbeat.php

Co-authored-by: Jeremy Herve <jeremy@jeremy.hu>

leogermani committed:
add deprecation notices

leogermani commented:
@jeherve As I feared tests have broken after adding the deprecated notices...

I took a path of letting the class fully backward compatible and not follow the rabbit hole of replacing all calls it... What do you think? Should we remove the notices or is it better to clean all the usage of these methods?

jeherve commented:
> What do you think? Should we remove the notices or is it better to clean all the usage of these methods?

Maybe now is a good time to clean things up then?

leogermani committed:
Fix test Jetpack_Heartbeat class

leogermani committed:
add deprecated notice  to cron_exec

leogermani commented:
@jeherve Fixed tests. Check https://github.com/Automattic/jetpack/pull/16285/commits/47dac496b797d86086316c8de1dbff43656eb9a2 and https://github.com/Automattic/jetpack/pull/16285/commits/1addc07e2c93b1c3d05b5b1a2221dd694e51213d

I think we can do some further cleaning, for example making sync use the new class directly and drop the usage  of deprecated code, but I'd rather do this in a separated PR. I've created an issue for that: https://github.com/Automattic/jetpack/issues/16789

leogermani commented:
@jeherve quick ping to move this forward

jeherve reviewed (approved):
Looking good now 🚢
